modules/dnn/src/tengine4dnn/src/tengine_graph_convolution.cpp

```diff
@@ -0,0 +1,363 @@
+        }
+    }
+
+    if (!ok)
+    {
+        destroy_graph(graph);
+        return nullptr;
+    }
+    return graph;
+}
+
+void tengine_set_Winograd(bool flag)
+{
+    // Winograd on/off .
+    setenv("NO_WINO", "1", (int)!flag);
+}
+
+bool tengine_forward(float *input_, int inch, int group, int in_h, int in_w,
+                        float *output_, int out_b, int outch, int out_h, int out_w,
+                        float *kernel_, int kernel_s ,int kernel_h, int kernel_w,
+                        float *teg_bias, int stride_h,int stride_w,
+                        int pad_h, int pad_w,  int dilation_h, int dilation_w,
+                        size_t wstep, std::string padMode)
+{
+    graph_t graph = NULL;
+    std::vector<float> teg_weight_vec;
+    float *teg_weight = NULL;
+    int kernel_inwh = (inch / group) * kernel_w * kernel_h;
+    // Do not using the activation fuse mode, just convolution only.
+    int activation = -1;
+
+    if (!(kernel_s == 2 && kernel_h == kernel_w && pad_h == pad_w
+        && dilation_h == dilation_w && stride_h == stride_w
+        && out_b == 1 && pad_h < 10)) // just for Conv2D
+        return false;
+
+    {
+        /*printf("Tengine: input (1 x %d x %d x %d),output (%d x %d x %d x %d), kernel (%d x %d), stride (%d x %d), dilation (%d x %d), pad (%d x %d).\n",
+               inch, in_h, in_w,
+               out_b,outch,out_h,out_w,
+               kernel_w, kernel_h,
+               stride_w, stride_h,
+               dilation_w, dilation_h,
+               pad_w,pad_h);*/
+        tengine_set_Winograd(false);    // Default not using the winograd algorithm.
+
+        // weight
+        if (kernel_inwh != wstep)
+        {
+            teg_weight_vec.resize(kernel_inwh * outch);
+            teg_weight = &teg_weight_vec[0];
+            for (int i=0; i<outch; i++)
+            {
+                memcpy(teg_weight+i*kernel_inwh, kernel_+i*wstep, kernel_inwh*FLOAT_TO_REALSIZE);
+            }
+        }
+        else
+        {
+            teg_weight = kernel_;
+        }
+
+        /* initial the resoruce of tengine */
+        init_tengine();
+
+        /* create the convolution graph */
+        graph = create_conv_graph( input_, inch, group, in_h, in_w,
+                                    output_, outch, out_h, out_w,
+                                    kernel_h, kernel_w, stride_h,stride_w,
+                                    pad_h, pad_w, dilation_h, dilation_w, activation,
+                                    teg_weight , teg_bias , padMode);
+
+        /* prerun */
+        if(prerun_graph(graph) < 0)
+        {
+            CV_LOG_WARNING(NULL, "Tengine :prerun_graph failed .");
+            return false ;
+        }
+
+        /* run */
+        if(run_graph(graph, 1) < 0)
+        {
+            CV_LOG_WARNING(NULL,"Tengine :run_graph failed .");
+            return false ;
+        }
+
+        postrun_graph(graph);
+        destroy_graph(graph);
```

Comment:
Is that optimal to create and destroy graph every run?

Reply:
Yes, it's not very optimal. I guess, each graph consumes some memory and there is no way currently to make it to use external buffers. If the graph is constructed once, the Tengine-accelerated OpenCV DNN will consume much more memory. For now the performance increase because of using Tengine just for convolution outweights the overhead of the graph construction. I guess, some better solution will be developed later